packages/flutter/lib/src/widgets/image_filter.dart

```diff
@@ -9,12 +9,17 @@ import 'package:flutter/rendering.dart';
 
 import 'framework.dart';
 
+/// Signature for a function that creates an [ImageFilter] for a given [Rect].
+///
+/// Used by [RenderImageFiltered] and the [ImageFiltered] widget.
+typedef ImageFilterCallback = ImageFilter Function(Rect bounds);
```

Comment:
this should take a Size, not a Rect, because we cannot guarantee that we'll call it again when the position of the rect changes (the topLeft of the rect, i.e. the offset passed to paint, can vary in ways that have little to do with the final rendering, e.g. based on what ancestor render objects create new layers)

Reply:
I'm confused as to how Transform works now. It basically takes the transform provided by the developer, which is relative to the bounds of the child, and biases it by the offset parameter from the call to paint() and the transform would not work if it could not trust that offset. The bias is the standard `..translate(offset)..concat(transform)..translate(-offset)` that repositions the origin "anchor" to which the transform applies.

So, if you can't use the offset in this way, then how does Transform work?

Reply:
No changes here in the latest commit. It seems to work and I'm not sure how your warnings are compatible with the way that RenderTransform, and particularly its associated layer, are written.

Reply:
@Hixie Perhaps I should clarify. What is important here is that the final coordinate space of any RenderObject is composed of 2 parts. The transform in the context and then an offset in the paint() call. The offset is there to avoid having a lot of individual translate and restore calls on the transform. Ideally we could get rid of offset entirely if every enclosing parent would have translated by the offset and recursed rather than just updating the offset.

The transform in a matrix filter needs to be relative to the context transform. If all parents worked by having used translate on the transform instead of updating the offset, then it would only need the size (and in many cases the app already knew the size of many objects anyway).

If a parent added a new layer, that doesn't affect the transform. Sure, when we render its children to the temporary texture in the engine we need to deal with the fact that the children are no longer rendering to the entire surface and are only rendering to a tiny temporary texture, but we do so by updating the transform under which they render. From their perspective, they are still rendering within the coordinate system that their RenderObject accepted and all of the fudging of the transform to make this appear to be true is transparent to the numbers they use in their rendering calls.

And none of that fudging matters here. What matters is that after the layout pass is done, this widget and its RenderObject should draw with numbers that go from LT => RB. If the LT changes with respect to the nearest ancestor that modified the context transform, then a widget/RO absolutely must be informed of the new offset. Consider a basic component that just draws its background. It draws it from LT => RB and those numbers are baked into the Picture that is created. That Picture can't be used again if there is a new LT, the rendering calls must be resubmitted with new numbers. So, I don't understand how the offset can be modified without rendering all of the children. Here is the code in _RenderColorBox that draws the background of a simple container. Note that it uses the offset in expressing its drawRect, so the RenderObject.paint call must be called again if the offset were to ever change:

(From [basic.dart](https://github.com/flutter/flutter/blob/f70ad37ec9df48d7ad6f96c304a25c867f21ada0/packages/flutter/lib/src/widgets/basic.dart#L7691))
```
      context.canvas.drawRect(offset & size, Paint()..color = color);
```

Am I missing something?